Ubuntu: https://bugs.launchpad.net/ubuntu/+source/grub2/+bug/671097
Upstream: not really applicable
Description: Explicitly ignore linux kernels with names ending in -ec2.
 These ubuntu kernels are intended only for Amazon EC2 guests where grub2
 will not ever be the loader.

--- grub2/util/grub.d/10_linux.in	2010-11-04 16:15:51.000000000 -0400
+++ grub2.new/util/grub.d/10_linux.in	2010-11-04 16:16:00.288879090 -0400
@@ -110,8 +110,20 @@
 EOF
 }
 
+blacklisted_kernel() {
+  local r=""
+  case "${1}" in
+    *-ec2) r="linux-ec2";; # LP: #671097
+  esac
+  [ -n "${r}" ] && { echo "Skipping linux image [$r]: ${1}" >&2; return 0; }
+  return 1;
+}
+
 list=`for i in /boot/vmlinu[xz]-* /vmlinu[xz]-* ; do
-        if grub_file_is_not_garbage "$i" ; then echo -n "$i " ; fi
+        if grub_file_is_not_garbage "$i" &&
+           ! blacklisted_kernel "$i"; then
+          echo -n "$i ";
+        fi
       done`
 prepare_boot_cache=
 
